django/core/views.py

```diff
@@ -358,109 +265,310 @@ def discourse_sso(request):
+
 
-    context_object_name = "object"
+class ToggleFollowUser(APIView):
+    permission_classes = (IsAuthenticated,)
+
+    def post(self, request, *args, **kwargs):
+        logger.debug("POST with request data: %s", request.data)
+        username = request.data["username"]
+        source = request.user
+        target = User.objects.get(username=username)
+        follow_user, created = FollowUser.objects.get_or_create(
+            source=source, target=target
+        )
+        if created:
+            target.following.add(follow_user)
+        else:
+            follow_user.delete()
+        return Response({"following": created})
+
+
+class TagListView(mixins.ListModelMixin, viewsets.GenericViewSet):
+    serializer_class = TagSerializer
+    queryset = Tag.objects.all()
+    pagination_class = SmallResultSetPagination
+    permission_classes = (AllowAny,)
+
+    def get_queryset(self):
+        query = self.request.query_params.get("query")
+        queryset = Tag.objects.all()
+        if query:
+            queryset = queryset.filter(name__icontains=query)
+        return queryset.order_by("name")
+
+
+class MemberProfileFilter(filters.BaseFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        if view.action != "list":
+            return queryset
+        query_params = request.query_params
+        qs = query_params.get("query")
+        tags = query_params.getlist("tags")
+        return get_search_queryset(qs, queryset, tags=tags)
+
+
+class MemberProfileViewSet(CommonViewSetMixin, HtmlNoDeleteViewSet):
+    lookup_field = "user__pk"
+    lookup_url_kwarg = "pk"
+    queryset = MemberProfile.objects.public().with_tags()
+    pagination_class = SmallResultSetPagination
+    filter_backends = (MemberProfileFilter,)
+    permission_classes = (ObjectPermissions,)
+    context_object_name = "profile"
+    context_list_name = "profiles"
+
+    def get_serializer_class(self):
+        if self.action == "list":
+            return RelatedMemberProfileSerializer
+        else:
+            return MemberProfileSerializer
+
+    def get_queryset(self):
+        if self.action == "retrieve":
+            return self.queryset.with_peer_review_invitations()
+        else:
+            return self.queryset.with_user()
 
     def get_retrieve_context(self, instance):
-        context = {self.context_object_name: instance}
+        context = super().get_retrieve_context(instance)
+        accessing_user = self.request.user
+        # FIXME: ideally this functionality should be in the library app, though its not clear
+        # whether this can be done without a much more complicated workaround or a major
+        # re-organization of apps
+        logger.debug("Finding models for user %s", instance.user)
+        context["codebases"] = (
+            Codebase.objects.accessible(accessing_user)
+            .filter_by_contributor(instance.user)
+            .with_tags()
+            .with_featured_images()
```

Comment:
This is the only bit of weirdness left (importing from `library` in `core`). Couldn't come up with a clean solution for it but there must be one..

Reply:
I think this about as best as we can make it for balancing simplicity + understandability

In the future if Django makes it easier to move models across apps it might be better for `core.MemberProfile` and its associated views / urls / templates to move to `home.MemberProfile` so that the `core` doesn't have any dependencies on other apps, while allowing other apps to have side linkages with each other as necessary